src/engine/database/mod.rs

```diff
@@ -134,6 +135,86 @@ pub trait DatabaseEndpoint {
     async fn pool(&self) -> Result<DatabasePool, Error>;
 }
 
+/// Represents the different type of comparison match operations
+#[derive(Clone, Debug)]
+pub enum Operation {
+    EQ,
+    CONTAINS,
+    IN,
+    GT,
+    GTE,
+    LT,
+    LTE
+}
+
+/// Struct representing a value comparison. In query operations, visitors take provided 
+/// operation/value nested map and converted them into a `Comparison` struct and pass
+/// it on the database-specific transaction for use in creating match portion of queries. 
+#[derive(Clone, Debug)]
+pub struct Comparison {
+    operation: Operation,
+    operand: Value,
+    negated: bool
+}
+
+impl Comparison {
+
+    pub fn new(operation: Operation, negated: bool, operand: Value) -> Self {
+        Comparison {
+            operation,
+            operand,
+            negated
+        }
+    }
+
+    pub fn default(v: Value) -> Self {
+        Self::new(Operation::EQ, false, v)
+    }
+
+}
+
+impl TryFrom<Value> for Comparison {
+
+    type Error = Error;
+
+    fn try_from(v: Value) -> Result<Comparison, Error> {
+        Ok(match v {
+            Value::String(_) => Comparison::default(v),
+            Value::Int64(_) => Comparison::default(v),
+            Value::Float64(_) => Comparison::default(v),
+            Value::Bool(_) => Comparison::default(v),
+            Value::Map(mut m) => {
+                let operation_str = m.keys().next()
+                    .ok_or(Error::ComparisonParsingFailed { message: "Comparison keys empty".to_string() })?
+                    .clone(); // TODO: is there a way to not clone here?
+                let operand = m.remove(&operation_str).unwrap();
+                Comparison::new(
+                    match operation_str.as_ref() {
+                        "EQ" => Operation::EQ,
+                        "NOTEQ" => Operation::EQ,
+                        "CONTAINS" => Operation::CONTAINS,
+                        "NOTCONTAINS" => Operation::CONTAINS,
+                        "IN" => Operation::IN,
+                        "NOTIN" => Operation::IN,
+                        "GT" => Operation::GT,
+                        "GTE" => Operation::GTE,
+                        "LT" => Operation::LT,
+                        "LTE" => Operation::LTE,
+                        _ => return Err(Error::ComparisonParsingFailed { 
+                            message: format!("Unknown operation {}", operation_str) 
+                        })
+                    },
+                    matches!(operation_str.as_ref(), "NOTEQ" | "NOTCONTAINS" | "NOTIN"),
+                    operand
+                )
+            },
+            _ => {
+                return Err(Error::ComparisonParsingFailed { message: format!("Unsupported Value type: {:#?}",  v)})
```

Comment:
Could this error be captured under the existing TypeNotExpected error? One of the examples it gives is a `Value` with an unexpected `enum` variant.  I'm looking to avoid adding additional error variants unless we really need them, but definitely push back if I'm going too far, there.

Reply:
I agree on wanting to keep error variants to a minimum. I did consider `TypeNotExpected`, but it has not fields and I wanted the ability to give some context on what was happening in case we ran into issues on what is a new and significant change. I'm totally onboard with re-using `TypeNotExpected` if it has a details/message field. What do you think makes more sense?

Reply:
Adding a message field to the `TypeNotExpected` variant so that it better suits your purpose sounds good to me.